text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+
+Note that despite the name, `struct` expressions are not limited to `struct`s. They are used to
+initialize `enum` variants and `union`s as well. For `enum`s and `union`s, this restriction only
+applies to the specific variant being constructed. For example, the following is allowed:
+
+```rust
+pub enum Foo {
+    Alpha { mut(crate) x: u8 },
+    Beta { y: u8 },
+}
+
+// In another crate:
+Foo::Beta { y: 5 };
+```
+
+In this example, `Foo::Alpha { x: 5 }` is allowed when it is in the same crate as `Foo`. This is
+because `x` is not restricted within this scope, so the field can be freely mutated. Because of
+this, the previous concern about upholding variants is not applicable.
+
+## `mod` as a restriction scope
+
+`pub` can be restricted in a few ways. The most common is `pub(crate)`, which restricts visibility
+to the current crate. However, there are other ways to restrict visibility. You can also use
+`pub(super)` to restrict visibility to the parent module, and `pub(in path)` to restrict visibility
+to `path`, as long as that path is an ancestor of the location it is used. There is one additional
+case you have likely never encountered: `pub(self)`. The reason you have likely never seen this
+before is that it is completely useless. Why? `pub(self)` is identical to private, which is the
+default visibility.
+
+While `pub(self)` should never be written in ordinary code, `impl(self)` and `mut(self)` are quite
+different. As the default restriction is unrestricted implementation and unrestricted mutation, it
+is more than likely that `impl(self)` and `mut(self)` will be quite common. During previous
+discussions about syntax, the unclear meaning of `impl(self)` and `mut(self)` was brought up. Syntax
+should have a clear meaning, and `impl(self)` and `mut(self)` are not clear. While they should be
+allowed for consistency with `pub(self)`, an alternative was needed with the same behavior.
+`impl(mod)` and `mut(mod)` should have a sufficiently clear meaning: implementing the trait and
+mutating the field are restricted to the current module. `impl(in mod)` and `mut(in mod)` are
+accepted as well. The behavior of `impl(mod)` is identical to `impl(self)`; likewise for `mut(mod)`
+and `mut(self)`.
+
+For consistency with the new restriction syntax, `pub(mod)` and `pub(in mod)` are also allowed.
+
+# Reference-level explanation
+
+## Syntax
+
+Using the syntax from [the reference for `struct`s][struct syntax], the change needed to support
+`mut` restrictions is quite small.
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   IDENTIFIER : Type
+
+TupleField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   Type
+
++MutRestriction :
++     mut
++   | mut ( crate )
++   | mut ( self )
++   | mut ( mod )
++   | mut ( super )
++   | mut ( in SimplePath )
++   | mut ( in mod )
```

Comment:
imho `mut(in mod)` is redundant with `mut(mod)` and is not justified by the existing `pub(in <path>)`. `pub(in self)` works only because `self` is a path, `mod` is not (unless you're proposing `mod` should become a path?)

Reply:
I don't think we should add this, or `mod`, at all.